Add a workspace list to the bar

swfrench opened the pull request:
By adding '+L' to the `bar_format`, you can now get a list of active
workspaces associated with the current screen (where active is defined
as either (a) containing windows or (b) being associated with the
current region, regardless of whether windows are present).

Two '*' characters bracketing a workspace indicates it is the workspace
of the current region.

For example, suppose we have windows on workspaces 1 and 5, and we are
currently on workspace 4. Further, suppose workspace 1 is called "web".
Then, the list would look like: `[ 1:web *4* 5 ]`

This is implemented in `bar_workspace_list` in spectrwm.c.

marcopeereboom commented:
I am ok with this.

swfrench commented:
Great! Doesn't look like I need to rebase, but let me know if there's anything you would like me to do.

mvhulten commented:
The window manager i3 shows such a list.  Right now I press M-0,9,8,... before quiting spectrwm, just to check if I forgot about any open windows.  I would be a happy to see this useful feature in spectrwm.

Could someone with write access merge the pull request?

marcopeereboom commented:
@LordReg what are yor thoughts?

LordReg commented:
@mvhulten ws_next/ws_prev (default: MOD-Right/MOD-Left) is the quick and easy way to check for non-empty workspaces.

@marcopeereboom The concept of having a list of 'workspaces with windows' is good but I want to see a more open-ended design. We don't need another +U style 'workspace list' that only covers a specific case.  Using both +U and +L could eat the whole bar!
What I envision is a general 'workspace overview indicator' that is minimal and potentially configurable to indicate some combination of:

- workspaces with windows
- workspaces with urgent windows
- named workspaces (including empty)
- all workspaces
- etc

I could see myself using an overview indicator in place of +D, +U and possibly +I.

marcopeereboom commented:
Since I find this all silly @swfrench and @mvhulten need to please @LordReg.  His ok is mine as well.

swfrench commented:
Thanks for taking a look!

@LordReg - Sure, I'd be happy to try for something more general. Currently, the proposed '+L' includes roughly the same information as +D and +I (well, a superset since it shows +I:+D for all workspaces with windows, not just current). What it clearly lacks is integration with urgency hints, but this should be simple.

I'll think a bit on how I could merge together the various example use cases you described in a way that should be straightforward to configure.

swfrench commented:
Many thanks for your patience while it's taken me a while to get back to this.

I've updated the proposed workspace list feature to integrate some of your feedback, @LordReg.

First, it now supports multiple subsets of listed workspaces:
* active (i.e. has windows or is associated with the current region - the default);
* active + named (i.e. in addition to active, show any named workspace); or
* all (i.e. show all workspaces).

The current workspace is still annotated with a '*' character preceding. The set is controlled via the `bar_ws_list_show` configuration key. Acceptable values are `active`, `named`, and `all`, which can be expanded to accommodate other subsets implemented later on.

In addition, the list can now display an urgency hint (a '!' character) next to any workspace with an urgent window. Thus, it will no longer be necessary to use the space-hogging combination of +U and +L in order to get per-workspace urgency hints. This is enabled via the `bar_ws_list_urgent` configuration key.

Revised example from the commit message: Suppose we have windows on workspaces 1 and 5, and we are currently on workspace 4. Further, suppose workspace 1 is called "web" and workspace 5 contains an urgent window (and `bar_ws_list_urgent` is enabled). Then, the list would look like:

    [ 1:web *4  !5 ]

I've rebased the PR branch onto the latest master and squashed the revisions into a single commit. Let me know if you have any further thoughts.

mvhulten commented:
This sounds perfect, Scott!

dajohi commented on spectrwm.c:
```diff
@@ -8958,6 +9025,19 @@ setconfvalue(const char *selector, const char *value, int flags)
+		else if (strcmp(value, "all") == 0)
+			bar_ws_list_show = SWM_BAR_WS_LIST_ALL;
+		else
+			errx(1, "invalid bar_ws_list_show");
```
should this blow up? or just show an error?

swfrench replied:
Good question. I was basically treating this with similar severity as invalid values for other bar-related configs (e.g. justification).

If it would be preferable to adopt a warning in the bar - potentially in addition to a sensible default for the set of listed workspaces (e.g. `SWM_BAR_WS_LIST_ACTIVE`) - I'd certainly be open to that.

mvhulten replied:
I prefer a warning in the bar.  Then I can fix and M-q spectrwm quickly.

For the rest (I just tried it on OpenBSD-current), @swfrench's code works fine.  I like `active` best.  Urgency windows also prefix with `!`.  All useful features!

swfrench replied:
Thanks for the follow-up! I'd be happy to swap out the errx for a warning in the bar. That said, it looks like @LordReg has put together an alternative impl of +L offering more fine-grained control over the included set of workspaces (rather than the three fixed flavors here).

LordReg commented:
This is what I was thinking: [workspace_indicator branch](/conformal/spectrwm/tree/workspace_indicator)
You can configure what is displayed in the workspace indicator with a comma-separated list of options.
Default is `workspace_indicator = listcurrent,listactive,markcurrent,printnames`

swfrench commented:
I was just testing my PR after rebasing against upstream and saw the branch you mention here. Nice!

I'm happy to revert this if your implementation of +L is preferred. Seems that it covers a superset of the fixed configuration flavors offered in mine, which is great.
